### PR TITLE
[SR-3811]Optimize Count Distinct using prefetch technique

### PR DESCRIPTION
**Pull request by dirtysalt**
ref: #139 

1. rename phmap methods 
2. prefetch in distinct agg state

### CODE REVIEW

File: be/src/exprs/agg/distinct.h [hunk 1]
```diff
@@ -249,23 +271,56 @@ class TDistinctAggregateFunction
...
-                if ((i + prefetch_dist) < batch_size) {
-                    agg_state.prefetch(data[i + prefetch_dist]);
-                }
-                mem_usage += agg_state.update(data[i]);
+            mem_usage += agg_state.update_with_hash(mem_pool, container_data[i], cache[i].hash_value);
+        }
+        ctx->impl()->add_mem_usage(mem_usage);
+    }
+
+    void update_batch(FunctionContext* ctx, size_t batch_size, size_t state_offset, const Column** columns,
```
Comment (imay) [inline be/src/exprs/agg/distinct.h:297]: Better to add comment to explain why split this update operation into two loop.
And why use prefetch.

**Comment by dirtysalt** [comment]
passed daily test

File: be/src/exprs/agg/distinct.h (hunk 1 above)
Comment (dirtysalt) [reply be/src/exprs/agg/distinct.h:301]: comments added.

**Review by satanson — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

File: be/src/exprs/agg/distinct.h [hunk 2]
```diff
@@ -243,29 +265,70 @@ class TDistinctAggregateFunction
...
+        std::vector<CacheEntry> cache(batch_size);
+        const auto& container_data = column->get_data();
+        for (size_t i = 0; i < batch_size; ++i) {
+            AggDataPtr state = states[i] + state_offset;
+            auto& agg_state = this->data(state);
+            size_t hash_value = agg_state.set.hash_function()(container_data[i]);
+            cache[i] = CacheEntry{&agg_state, hash_value};
+        }
+        // This is just an empirical value based on benchmark, and you can tweak it if more proper value is found.
+        size_t prefetch_index = 16;
```
Comment (starrocks-xueli) [inline be/src/exprs/agg/distinct.h:323]: can you use the macro in last pr instead of 16?

File: be/src/util/phmap/phmap.h [hunk 3]
```diff
@@ -1702,7 +1702,7 @@ class raw_hash_set {
     };
 
     template <class K, class... Args>
-    std::pair<iterator, bool> emplace_decomposable(const K& key, size_t hashval, Args&&... args) {
+    std::pair<iterator, bool> emplace_decomposable_with_hash(const K& key, size_t hashval, Args&&... args) {
```
Comment (starrocks-xueli) [inline be/src/util/phmap/phmap.h:1705]: why change the name?

**Pull request merged**